recipes/lora_finetune_single_device.py

```diff
@@ -491,20 +515,54 @@ def save_checkpoint(self, epoch: int) -> None:
             )
 
         # Move to CPU to avoid a copy on GPU
-        state_dict = {k: v.cpu() for k, v in self._model.state_dict().items()}
+        # state_dict = {k: v.cpu() for k, v in self._model.state_dict().items()}
+
+        # === Create a FakeTensor version of the state_dict (preserves storage aliasing) ===
+        state_dict = self._model.state_dict()
+        mode = FakeTensorMode()
+        converter = FakeTensorConverter()
+        fake_state_dict = OrderedDict()
+        for k, v in state_dict.items():
+            # remove adapters (we don't want them in merged_state_dict)
+            if "lora_a" in k or "lora_b" in k:
+                continue
+            if isinstance(v, NF4Tensor):
+                fake_state_dict[k] = converter.from_real_tensor(mode, v).to(torch.bfloat16).cpu()
+            else:
+                fake_state_dict[k] = converter.from_real_tensor(mode, v).cpu()
+
+        # === Save and load fake state_dict ===
+        # dest_state_dict is now backed by an mmap-ed file that can be written to
+        dest_state_dict_path = "/tmp/llama_state_dict.pt"
+        torch.save(fake_state_dict, dest_state_dict_path)
+        torch.serialization.set_default_mmap_options(MAP_SHARED)
+        dest_state_dict = torch.load(dest_state_dict_path, mmap=True, weights_only=True)
+        torch.serialization.set_default_mmap_options(MAP_PRIVATE)
+
+        # This bit does D2H one by one and since dest_state_dict is backed by mmap --> won't OOM :)
+        for k in state_dict.keys():
+            if 'lora_a' in k or 'lora_b' in k:
+                # adapters are very small ~0.04GB so we don't need to use mmap when offloading to CPU
+                dest_state_dict[k] = state_dict[k].cpu()
+            if isinstance(state_dict[k], NF4Tensor):
+                dest_state_dict[k].copy_(state_dict[k].to(torch.bfloat16))
+            else:
+                dest_state_dict[k].copy_(state_dict[k])
 
         # Construct the full state dict with LoRA weights merged into base LLM weights
         merged_state_dict = get_merged_lora_ckpt(
             state_dict,
             rank=self._lora_rank,
             alpha=self._lora_alpha,
+            dest_state_dict=dest_state_dict,
```

Comment:
(another noob q:) so after this, we can save the checkpoint normally (e.g. in the call to `save_checkpoint` on L580), even though we are now saving something like `{"model": dest_state_dict, ... (other stuff)}` to a separate file from the mmapped one?

Reply:
We should think of dest_state_dict as being "mmap-backed"

The final `save_checkpoint` on L580 is resaving a new checkpoint with `{"model": dest_state_dict, ... (other stuff)}` to a new file

We could potentially do something smarter than what we're doing right now but the code changes would be more invasive than they are now, given that the checkpointers seem to do some remapping though I refrained from doing this as a v1, wdyt?